[mlir][IR] Add builtin `TokenType` (#195640)
llvm/llvm-project: mlir/docs/Tokens.md

As with regular use-def chains, a token def-use chain is unidirectional. A
token use points to the token's definition and not the other way around.
Transformations can remove the use of a token without having to touch or
inspect the definition of the token.

## Structural Contract

A token use cannot be substituted with another token value: the use of a token
points directly to a specific producer. Generic transformations must not alter
or break this link. New uses of a token can be introduced safely. As a
consequence:

1. A token must not appear as a forwarded value, e.g.:
    * a forwarded result/operand of a `CallOpInterface` op,
    * an argument or result type of a `FunctionOpInterface` op (a token
      block argument *inside* a function body is fine — what is disallowed
      is forwarding tokens across the call/return boundary),
    * a successor operand or successor block argument of a
      `BranchOpInterface` op,
    * a forwarded operand to/from any region of a `RegionBranchOpInterface`
      op (iter-args, region results, yielded values), or
    * the result of any op that selects or merges values it does not

>       understand (e.g. `arith.select`).

Reviewer: Is this gonna be verified? enforced? Is it planned in this PR or follow up? At the moment only enforcement is through AnyType, but one does not need to use that in these cases.

Reviewer 2: Can you describe the failure mode you have in mind?

Reviewer: It is unclear to me how e.g. "A token must not appear as a forwarded result/operand of a `CallOpInterface` op" is going to be enforced?

Nothing forces the of use `AnyType` for these interfaces , e.g. `LLVM_CallOp` declares `CallOpInterface` but its operands are `Variadic<LLVM_Type>:$callee_operands,` where `LLVM_Type` does not "inherit" from `AnyType`, therefore does not check for token-type anyhow.

The same applies for other interfaces.

Reviewer 2: > It is unclear to me how e.g. "A token must not appear as a forwarded result/operand of a CallOpInterface op" is going to be enforced?

It won't in the current implementation (Matthias is looking into the traits verification, but let's talk about the current implementation), but that shouldn't be an actual issue in reality:

>  e.g. LLVM_CallOp declares CallOpInterface but its operands are Variadic<LLVM_Type>:$callee_operands, where LLVM_Type does not "inherit" from AnyType, therefore does not check for token-type anyhow.

It does not need to check for token explicitly, because it already does a `if (llvm::isa<...>)` over a list of allowed types. This is why `AnyType` is the only one we need to update: all the other ODS predicates are already checking positively that the types belong to some constraints and would exclude the new `TokenType`.

Reviewer: I got that, but nothing technically prevents me createing `call` op that takes token as operand. So the LangRef constraints are only "soft" besides `AnyType` check which is the only proposed hard-constraint. 

I’m okay with the current proposal. It just wasn’t clear to me whether this was intentionally omitted, deliberately postponed, or simply overlooked. We maybe want to explicitly mention this “soft-correctness” limitation in LangRef?

Reviewer 2: > I got that, but nothing technically prevents me createing call op that takes token as operand. 

You would have to write your own op in C++ or your own ODS constraints, because `AnyType` would catch it.

As I mentioned elsewhere, folks can always creatively find ways to shoot themselves in the foot, our verifiers are pretty defensive but can't be bulletproof either.

> We maybe want to explicitly mention this “soft-correctness” limitation in LangRef?

I don't think so: the rules aren't "softer" because we don't enforce them at some point, the implementation of the verifier and its limitation are to me disconnected from the "specification" that LangRef is meant to encode. 

In any case, the traits-solution will catch it.

Reviewer 2: Stay tuned, but resolving this.

2. Given a use of a token SSA value, its definition is guaranteed to be the

>    semantic producer of the token.

Reviewer: How is this supposed to be interpreted in regards to CSE? Can CSE merge two token producers? Are they semantic producers?
Maybe worth noting that tokens are nto CSE-stable (at least this is my intuition)?

Reviewer 2: > Can CSE merge two token producers?

I would say only if it merges the users. But I'm not sure the current text is clear on this. The original working document went into more details into what could be a "token identity" and the impact on OperationEquivalence for example. We're missing some of it right now.

Reviewer: Yeah, I believe @matthias-springer was advocating for and there was no real oposition to it, citation from RFC:

_5.3. Defined by Op Semantics: Drop the “5. Token is unique” property entirely and leave it up to the token-producing op. E.g., if CSE’ing of token-producing ops is desirable in your dialect, mark the op as pure. If not, leave the side effects unspecified (or add a AllocateToken effect), or utilize an IR design pattern that cannot be CSE’d (e.g., tokens are always block arguments)._

I guess, this property should me mentioned somewhere in `Tokens.md`.

Author: I had that spelled out explicitly in `Tokens.md` [in an earlier version](https://github.com/llvm/llvm-project/commit/86ba661da2f360bb7e51c85c17ecd4f62c4aa045#diff-04c04d9d6c774a0713d8a61f8618d1ab64430eb3f69be59399cf1db92bcbccbdR30).

```
Token-producing and token-consuming ops are subject to standard transformations
such as CSE, DCE and hoisting. If such transformations are not desirable due to
op semantics, common IR design patterns can be employed.
```

It's not really a property of the token type itself (but of the token-producing op), so I removed it. I could bring that part back. (The original wording may have been confusing, I'm not sure how to improve it. See discussion [here](https://github.com/llvm/llvm-project/pull/195640#discussion_r3225218825) and [here](https://github.com/llvm/llvm-project/pull/195640#discussion_r3225224045).)

> Can CSE merge two token producers?

The presence of tokens does not affect the CSE machinery. So my answer is "yes".

Whether CSE'ing is desirable/correct or not may depend on the semantics of token-producing ops. If necessary, you can design your token-producing ops in such a way that CSE would not merge them. E.g., by using different "identifier" attributes, by working with side effects, or by not using token OpResults in the first place (but instead token block arguments).

Reviewer 2: I don't think I'm on the same line at the moment that we can just apply the usual rule. We have a long discussion on how OperationEquivalence and CSE need to change, and this is intrinsically tied to Token semantics. 

>  E.g., by using different "identifier" attributes, by working with side effects, or by not using token OpResults in the first place (but instead token block arguments).

I don't quite understand why we should leave this up to just the user having to hack around the system no providing the complete set of feature right now.

Reviewer 2: > Yeah, I believe @matthias-springer was advocating for and there was no real oposition to it

I apologize for not voicing this in the RFC thread, I actually did on the draft design doc:

> I'm concerned about the implications if we introduce side-effects: this is in general degrading things significantly. What about being able to CSE two ops who have each a region with a token producing op inside for example?

> I would say that a goal of tokens would be to mesh with the ecosystem, adding side effect would be a big hammer IMO. Ops With recursive side effects, Would no longer be side effect free

Reviewer 2: Oh actually this was carried in the "downside" in RFC, which seemed enough to me to disqualify this, which is why I didn't comment further:

> Downside: Side effects are describing things that happen outside of SSA dataflow. Introducing side-effects into the token modeling makes it no longer a pure SSA representation and defeats most of the usefulness of a new construct. Side-effects are costly and pessimize / deactivate most optimizations.

(5.3 was added later, but the side-effects comment seems to apply equally)

Author: I am questioning this part: `We have a long discussion on how OperationEquivalence and CSE need to change, and this is intrinsically tied to Token semantics.`

Why do we have to give the token type such a strong semantics (as opposed to the op semantics)? It limits the things that you can do with a token. What if as a dialect designer, you want your two `%0 = make_token` ops to be CSE'd?

Reviewer 3: Since I feel a bit responsible for the `OperationEquivalence` meandering, and really like this PR, I'd like to add that I agree with Matthias here. I do think tokens should rely on the op semantics only. I've come around to the viewpoint that special cases only set a bad precedent for an interoperable MLIR ecosystem.

A big reason for why I am now much more comfortable with this is that I realized [operation properties can model unitary ops very well already](https://discourse.llvm.org/t/bug-in-operationequivalence-breaks-cse-on-linalg-index/85773/51). That is, upstream does in fact already allow you to declare a `Pure` op that can't be CSE'd, in a surprisingly idiomatic way I'd say.

There is also "Pure means always hoistable", which I still haven't come around on. But I fail to see a good faith argument about why a token-producing op would be somehow different in that regard to one one that doesn't.

Author: I updated the "Structural Contract" section to explicitly allow CSE and DCE. The key phrase is "semantic producer". This is the last main item that we need to agree on. PTAL.

3. A token cannot constant-fold. No constant of token type exists.

> 

Reviewer 2: ```suggestion

4. Use of a token is side-effect free: a token user follows the usual `isTriviallyDead()` rules.

```

Author: Makes sense. Just to confirm: users can still have side effects. But there can be no side effects associated with a token operand.

Reviewer 2: Right!

Reviewer: Nice! :)

back from a use and say "this token came from *that* specific op". The exact
structural contract is specified in the
[LangRef section on tokens](LangRef.md#token-type).

A token is an SSA value that has the builtin token type. The token type is 
parameterless, opaque and carries no runtime data. Apart from the structural
contract specified in the LangRef, tokens are like any other SSA values.

## Design Rationale

The token type allows operations to refer to another operation without a new
parallel def-use system for operations. It reuses the existing def-use
machinery for SSA. It introduces no changes to the generic op syntax, the
bytecode infrastructure or core C++ APIs around `Operation`.

As with regular def-use chains, a token def-use chain is unidirectional. A
token use points to the token's definition and not the other way around.
Transformations can remove the use of a token without having to touch or
inspect the definition of the token.

Because tokens are SSA values, they cannot cross `IsolatedFromAbove` region
boundaries. This is intentional: it allows passes to process isolated regions
concurrently without racing on def-use chains. When a token-like dependency

> must cross such a boundary, use a symbol instead.

Reviewer 2: ```suggestion
concurrently without racing on def-use chains. When a token-like dependency
must cross such a boundary, another mechanism must be used (e.g. a symbolic
reference using an attribute).
```

Reviewer 2: "Symbol" is overly specific to the `Symbol` traits/interfaces.

> # Tokens

Reviewer 2: Why isn't this a section of LangRef?

Author: The entire `Tokens.md` or a just a part of it? The current wording in this file is a bit verbose (design rationale, IR example, detailed structural contract) compared to other parts of the LangRef.

Reviewer 2: Seems verbose indeed, either some of it could stay in there and be referenced for information from LangRef (as well as from the TokenType ODS description), or what can't be in LangRef could be in the TokenType ODS description (so it'll appear in the builtin dialect page).

Author: I added a one-paragraph summary to the LangRef, along with a link to this document.

[TOC]

## Overview

Intuitively, a *token* value is a pointer to an operation (via an OpResult)
or a pointer to a region (via an entry block argument). A token cannot be
forwarded: a token def-use chain cannot be obscured by ops with forwarding
semantics such as `arith.select` or `cf.br`. This allows you to always walk
back from a use and say "this token came from *that* specific op". 

A token is an SSA value that has the builtin token type. The token type is

> parameterless, opaque and prints as `token`. A token carries no runtime data.

Reviewer 4: nit: I'd swap the print with the next phrase.

```
The token type is parameterless, opaque and carries no runtime data. A token prints as `token`.
```

Apart from the structural contract below, tokens are like any other SSA values.

## Design Rationale

The token type allows operations to refer to another operation without a new
parallel def-use system for operations. The existing def-use machinery for SSA

> values can be reused. Moreover, no changes were needed for the generic op

Reviewer 4: nit: `It introduces no changes to the generic op ...`.

The doc must read as if it has happened already and is accepted as correct, so we don't need _arguments for it_.

syntax, the bytecode infrastructure and core C++ APIs around `Operation`.

As with regular use-def chains, a token def-use chain is unidirectional. A

> token use points to the token's definition. (But not the other way around.)

Reviewer 2: ```suggestion
token use points to the token's definition and not the other way around.
```

Transformations can remove the use of a token without having to touch or
inspect the definition of the token. (Whether such a transformation is correct

> depends on the semantics of the token-producing and token-consuming ops.)

Reviewer 2: The parenthesis invalidates the previous sentence, I don't get the point here... Either you're saying "it's OK" or you should say "it may not be OK so it shouldn't be done without having knowledge of the producer".

Author: I wrote this with the `tf_executor.NextIteration.Sink` example in mind, which has extra invariants, in addition to the [token structural contract](https://github.com/llvm/llvm-project/pull/195640#discussion_r3225253970): it assumes a 1:1 mapping between source/sink ops. The C++ API allows you to remove a `Sink` op, but such a transformation on its own would be incorrect.

There are two kinds of invalid transformations: (1) the ones that trigger an immediate C++ assertion when you attempt them, and (2) the ones that can be performed but then lead to invalid IR or a lowering failure.

Example 1: You cannot remove an operation that still has uses. The `Operation::erase` call would trigger an assertion.

Example 2: You can replace the `scf.yield` terminator of an `scf.for` with `tensor.yield` without triggering a C++ assertion, but the resulting IR would be incorrect / invalid.

Does this paragraph make sense to you if I replace "transformations" with "C++ API"? If not, I can also entirely drop it.

Reviewer 2: > I wrote this with the tf_executor.NextIteration.Sink example in mind, which has extra invariants, in addition to the https://github.com/llvm/llvm-project/pull/195640#discussion_r3225253970: it assumes a 1:1 mapping between source/sink ops.

This 1:1 isn't something we can provide with the token type. We should describe what the token type provides, and how we can manipulate it. So: can we DCE a token use that does not have side-effects?

Author: Yes, we can DCE it. What I wrote here has nothing to do with the contract / semantics of the token type itself. I removed this sentence and the paragraph below.

Token-producing and token-consuming ops are subject to standard transformations
such as CSE, DCE and hoisting. If such transformations are not desirable due to
op semantics, common IR design patterns can be employed. To give a few examples:
Terminators or ops with side effects are not CSE'd or DCE'd. Region block
arguments semantically belong to the enclosing op and are never CSE'd, DCE'd or

> hoisted. Non-speculatability may also prevent hoisting.

Reviewer 2: This paragraph is confusing to me in what it tries to say or what the implications are. As a reader I wouldn't know what conclusion to make out of this.

Author: I also wrote this with the `tf_executor.NextIteration.Sink` example in mind.

For a handwritten pass / pattern, you can make sure that it respects the semantics / invariants (1:1 mapping) of the source/sink op. But what about standard transformations such as CSE, DCE, hoisting, etc.? This paragraph says that you can work with terminators, side effects, etc. to make CSE, DCE, hoisting, etc. respect your op semantics / invariants.

To be fair, this is orthogonal to tokens. You may need some of these mechanisms anyway, e.g., to make sure that an op (even without tokens) is not hoisted from the `tf_executor.graph` region. I can remove this paragraph if it's more confusing than useful.

## Structural Contract

> 1. A token must not appear as a forwarded value, e.g.:

Reviewer 4: This should be protected via the `AnyType` not containing `TokenType`, right?

Author: Correct.

> 1. A token must not appear as a forwarded value, e.g.:

Reviewer 2: I feel this should be a consequence of what should be the structural contract # 1, which I'd phrase along the line of:

> A token use can't be substituted with another token value: the use of a token points directly to a specific producer. Generic transformations must not alter or break this link. New uses of a token can be introduced safely.

Almost everything falls from this.

Author: Done.

    * a forwarded result/operand of a `CallOpInterface` op,
    * an argument or result type of a `FunctionOpInterface` op (a token
      block argument *inside* a function body is fine — what is disallowed
      is forwarding tokens across the call/return boundary),
    * a successor operand or successor block argument of a
      `BranchOpInterface` op,
    * a forwarded operand to/from any region of a `RegionBranchOpInterface`
      op (iter-args, region results, yielded values), or
    * the result of any op that selects or merges values it does not
      understand (e.g. `arith.select`).

2. As a consequence of (1), given a use of a token SSA value, its definition is

>    guaranteed to be the semantic producer of the token.

Reviewer 5: If we only modify ODS, this is a non-enforceable structural contract for dowstreams. Nothing prevents users from misusing tokens by negligence or error/omission. For example, what if someone in a downstream has:
```td
def NonIntType : Type<CPred<"!type.isInteger()">;
```

This change will break the above constraint and potentially the contract.

Reviewer 2: I'm fine not going entirely out of our way to avoid people from completely shooting themselves in the foot. 
The kind of scenario that would both have someone use something like your `NonIntType`  and at the same time  an unintended transformation that would inject a token there is unlikely enough. More importantly: while this would survive the verifier, this would still fail the next lowering in all likeliness.

3. A token cannot constant-fold. No constant of token type exists.

These properties mirror what LLVM IR already documents for its own
[`token` type](https://llvm.org/docs/LangRef.html#token-type).

## ODS Integration

Tokens are excluded from the default `AnyType` predicate, so an op that has
not opted in cannot accept a token as an arbitrary operand or result. This
restriction prevents tokens from being accidentally passed as operands with
forwarding semantics.

Three predicates are provided in `CommonTypeConstraints.td`:

| Predicate          | Accepts                              | Use when …                                                            |
| ------------------ | ------------------------------------ | ----------------------------------------------------------------------|
| `AnyType`          | any non-token type                   | the default; matches the historical meaning of "any type" pre-tokens. |

> | `AnyTypeOrToken`   | any type, including tokens           | the op legitimately accepts arbitrary types (including tokens).       |

Reviewer 5: What is the point for `AnyTypeOrToken`? IMO I think it should be discouraged in general. If a downstream wants to add it, that's up to them, but I don't see why upstream needs/wants it.

Reviewer 2: We can leave it out until we find a use-case for it, not that it would hurt defining it for completeness purposes either.

Reviewer 4: If we make it clear that `AnyType` doesn't accept tokens (and I think we do, everywhere), then I agree we only need to add `AnyTypeOrToken` when needed.

Author: I removed `AnyTypeOrToken` for now. I see a use case in the LLVM dialect. Some [intrinsics accept tokens](https://llvm.org/docs/LangRef.html#llvm-call-preallocated-teardown-intrinsic) and you're calling these intrinsics with `llvm.call` (`call_intrinsic`?), which must accept tokens and regular SSA values.

| `Token`            | only the builtin `TokenType`         | the op specifically takes a token operand/result.                     |

Example:

```tablegen
def MyConsumeOp : MyDialect_Op<"consume"> {
  let arguments = (ins Token:$scope, AnyType:$value);
}
```

## Examples

> ### Rejected: tokens in `AnyType` positions

Reviewer 4: Perhaps one example that is valid?

Author: There is no good example that I can add at this point. I could add something like this, but that probably won't help much:
```mlir
llvm.func @coro_begin(%arg0: !llvm.ptr) {
  %zero = llvm.mlir.constant(0 : i32) : i32
  %null = llvm.mlir.zero : !llvm.ptr
  %token = llvm.intr.coro.id %zero, %null, %null, %null : (i32, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> token
  llvm.intr.coro.begin %token, %arg0 : (token, !llvm.ptr) -> !llvm.ptr
  llvm.return
}
```

I'd say let's expand this once we have early-exit based with tokens. That's easier to understand than these LLVM intrinsics.

Intuitively, a *token* value is a pointer to an operation (via an OpResult)
or a pointer to a region (via an entry block argument). A token cannot be
forwarded: a token def-use chain cannot be obscured by ops with forwarding
semantics such as `arith.select` or `cf.br`. This allows you to always walk
back from a use and say "this token came from *that* specific op". 

A token is an SSA value that has the builtin token type. The token type is 
parameterless, opaque and carries no runtime data. A token prints as `token`.
Apart from the structural contract below, tokens are like any other SSA values.

## Design Rationale

The token type allows operations to refer to another operation without a new
parallel def-use system for operations. It reuses the existing def-use
machinery for SSA. It introduces no changes to the generic op syntax, the
bytecode infrastructure or core C++ APIs around `Operation`.

As with regular use-def chains, a token def-use chain is unidirectional. A
token use points to the token's definition and not the other way around.
Transformations can remove the use of a token without having to touch or
inspect the definition of the token.

## Structural Contract

> A token use cannot be substituted with another token value: the use of a token

Author: I'd rephrase this paragraph as follows:
```
Given a use of a token SSA value, its definition is guaranteed to be the
semantic producer of the token. Generic transformations must preserve this
invariant: they may not introduce a forwarding step between a use and its
producer, nor retarget a use to a producer with different semantics. New
uses of a token can be introduced safely. As a consequence:
```

This would leave the door open for CSE'ing of token-producing ops. Point 2 could be dropped. Instead, we could mention explicitly that the presence of tokens has no effect on CSE, DCE or hoisting.

Another problem with the current: `must not alter
or break this link` may suggest that DCE'ing of consumers (Point 4) is not allowed.

invariant: they may not introduce a forwarding step between a use and its
producer, nor retarget a use to a producer with different semantics. New
uses of a token can be introduced safely. As a consequence:

1. A token must not appear as a forwarded value, e.g.:
    * a forwarded result/operand of a `CallOpInterface` op,
    * an argument or result type of a `FunctionOpInterface` op,
    * a successor operand or successor block argument of a
      `BranchOpInterface` op,
    * a forwarded operand to/from any region of a `RegionBranchOpInterface`
      op (iter-args, region results, yielded values), or
    * the result of any op that selects or merges values it does not
      understand (e.g. `arith.select`).

2. A token cannot constant-fold. No constant of token type exists.

3. The presence of tokens has no effect on standard transformations such as
   CSE, DCE or hoisting.

4. Use of a token is side-effect free: a token user follows the usual
   `isTriviallyDead()` rules.

These properties mirror what LLVM IR already documents for its own
[`token` type](https://llvm.org/docs/LangRef.html#token-type).

Operations must opt in to producing or consuming tokens with

> `TokenProducerTrait` and `TokenConsumerTrait`.

Reviewer 6: Are they allowed to be both at the same time?

Reviewer 2: Yes, I don't see a reason to forbid it.

Example:

```tablegen
def MyProduceOp : MyDialect_Op<"produce", [TokenProducerTrait]> {
  let results = (outs Token:$token);
}

def MyConsumeOp : MyDialect_Op<"consume", [TokenConsumerTrait]> {
  let arguments = (ins Token:$scope, AnyType:$value);
}
```

Region entry block arguments of `token` type are also token producers and
require the parent operation to define `TokenProducerTrait`. Token block
arguments in non-entry blocks are rejected.

## Examples

### Rejected: tokens in `AnyType` positions

`scf.yield` operands have forwarding semantics. A token cannot be yielded from
a branch or a loop.

```mlir
// error: 'scf.if' op result #0 must be variadic of any non-token type,
//        but got 'token'
%t = scf.if %cond -> token {
  %a = my.token.produce : token
  scf.yield %a : token
} else {
  %b = my.token.produce : token
  scf.yield %b : token
}
```

`scf.if`'s results are declared with `Variadic<AnyType>` and `scf.yield`'s
operands likewise use `AnyType`. Because `AnyType` excludes tokens, yielding
(or returning) a token through `scf.if` (or any other op that has not

> explicitly opted in) is rejected.

Reviewer 6: This reads like an op could opt into yielding a token, which contradicts tokens not being forwarded.

Author: I rephrased this sentence.
